File: java/hamming/src/main/java/Hamming.java
import java.util.stream.IntStream;

public class Hamming {
    public static Integer compute(String strand, String other) {
        if (strand.length() != other.length()) {
            throw new IllegalArgumentException("Strands do not have equal length");
        }

        Integer distance = IntStream.range(0, strand.length())
                .reduce(0, (count, index) -> {
                    if (strand.charAt(index) != other.charAt(index)) {
                        count++;
                    }
                    return count;
                });

        return distance;
    }
}

File: go/hamming/hamming.go
package hamming

import (
	"errors"
	"strings"
)

const testVersion = 5

func Distance(a, b string) (int, error) {
	if len(a) != len(b) {
		return 0, errors.New("strands must have equal length to calculate hamming distance")
	}

	var distance int
	aReader := strings.NewReader(a)
	bReader := strings.NewReader(b)

	for index := 0; index < len(a); index++ {
		nextA, _ := aReader.ReadByte()
		nextB, _ := bReader.ReadByte()
		if nextA != nextB {
			distance++
		}
	}

	return distance, nil
}
